lib/ecs_compose/compare.rb

```diff
@@ -8,15 +8,22 @@ module Compare
     def self.sort_recursively(val)
       case val
       when Array
-        val.sort_by do |item|
-          if item.instance_of?(Hash)
-            item.to_a.sort
-          elsif item.instance_of?(Array)
-            # for an array of arrays, just return 1
-            1
+        val.sort do |a, b|
+          a1 = if a.instance_of?(Hash)
+            a.to_a.sort
+          elsif a.instance_of?(Array)
+            a.sort
           else 
-            item
+            a
           end
+          b1 = if b.instance_of?(Hash)
+            b.to_a.sort
+          elsif b.instance_of?(Array)
+            b.sort
+          else 
+            b
+          end
+          a1 <=> b1 || 1
```

Comment:
`a1 <=> b1` is not an ordering, unfortunately.

An ordering should change sign when the arguments are swapped:

```
irb(main):004:0> 1 <=> 2
=> -1
irb(main):005:0> 2 <=> 1
=> 1
```

But we have:

```
irb(main):007:0> (nil <=> 1) || 1
=> 1
irb(main):008:0> (1 <=> nil) || 1
=> 1
```

This will produce bad weirdness:

```
irb(main):009:0> [nil, 1, nil, 2, nil].sort {|a,b| a <=> b || 1 }
=> [nil, nil, 1, 2, nil]
```